src/ClientHandle.cpp

```diff
@@ -184,7 +188,23 @@ void cClientHandle::Destroy(void)
 		{
 			player->StopEveryoneFromTargetingMe();
 			player->SetIsTicking(false);
-			world->RemovePlayer(player, true);
+
+			if (WasAddedToWorld)
+			{
+				// If ownership was transferred, our own smart pointer should be unset
+				ASSERT(!m_PlayerPtr);
+
+				m_PlayerPtr = world->RemovePlayer(*player, true);
+
+				// And RemovePlayer should have returned a valid smart pointer
+				ASSERT(m_PlayerPtr);
+			}
+			else
+			{
+				// If ownership was not transferred, our own smart pointer should be valid and RemovePlayer's should not
+				ASSERT(m_PlayerPtr);
+				ASSERT(!world->RemovePlayer(*player, true));
```

Comment:
I could add a debug-only function to ensure there are no stray references left in the world or chunk, or just remove the check.

The intention of decoupling the lifetime of the player and client handle would refactor the WasAddedToWorld hack away, so I do not expect this to stay for ever. It is, after all, just a shim to get smart pointers working.